File: test/agent.py
import sys
sys.path.insert(0, '../python')
import pyrl

import numpy




class QLearningAgent(pyrl.Agent):
    def __init__(self, env, hyperparameters):

        pyrl.Agent.__init__(self, env)

        self.hyperparameters = hyperparameters
        self.env = env

        self.state      = 0
        self.state_prev = self.state;

        self.action      = 0;
        self.action_prev = 0;

        self.actions_count = env.get_actions_count()
        self.states_count  = self.env.get_observation().size()
        print(self.actions_count, self.states_count)

        self.q_table = numpy.zeros((self.states_count, self.actions_count))

    def argmax(self, v):
        res = 0
        for i in range(0, len(v)):
            if v[i] > v[res]:
                res = i

        return res

    def select_action(self, v):
        action = self.argmax(v)
        epsilon = self.hyperparameters.epsilon
        if self.run_best_enabled():
            epsilon = self.hyperparameters.epsilon_best

        r =  numpy.random.uniform(0.0, 1.0)
        if r < epsilon:
            action = numpy.random.randint(0, self.actions_count)

        return action



    def main(self):
        self.state_prev = self.state
        self.state      = self.env.get_observation().argmax()

        self.action_prev    = self.action
        self.action         = self.select_action(self.q_table[self.state])

        reward = self.env.reward()

        gamma = self.hyperparameters.gamma
        alpha = self.hyperparameters.alpha

        d = reward + gamma*numpy.argmax(self.q_table[self.state]) - self.q_table[self.state_prev][self.action_prev]

        self.q_table[self.state_prev][self.action_prev]+= alpha*d

        self.env.action(self.action)

    def print_table(self):
        print(self.q_table)



class SarsaAgent(pyrl.Agent):
    def __init__(self, env, hyperparameters):

        pyrl.Agent.__init__(self, env)

        self.hyperparameters = hyperparameters
        self.env = env

        self.state      = 0
        self.state_prev = self.state;

        self.action      = 0;
        self.action_prev = 0;

        self.actions_count = env.get_actions_count()
        self.states_count  = self.env.get_observation().size()
        print(self.actions_count, self.states_count)

        self.q_table = numpy.zeros((self.states_count, self.actions_count))

    def argmax(self, v):
        res = 0
        for i in range(0, len(v)):
            if v[i] > v[res]:
                res = i

        return res

    def select_action(self, v):
        action = self.argmax(v)
        epsilon = self.hyperparameters.epsilon
        if self.run_best_enabled():
            epsilon = self.hyperparameters.epsilon_best

        r =  numpy.random.uniform(0.0, 1.0)
        if r < epsilon:
            action = numpy.random.randint(0, self.actions_count)

        return action


    def main(self):
        self.state_prev = self.state
        self.state      = self.env.get_observation().argmax()

        self.action_prev    = self.action
        self.action         = self.select_action(self.q_table[self.state])

        reward = self.env.reward()

        gamma = self.hyperparameters.gamma
        alpha = self.hyperparameters.alpha

        d = reward + gamma*self.q_table[self.state][self.action] - self.q_table[self.state_prev][self.action_prev]

        self.q_table[self.state_prev][self.action_prev]+= alpha*d

        self.env.action(self.action)

    def print_table(self):
        print(self.q_table)

File: libs_agent/agent.cpp
#include "agent.h"


Agent::Agent()
{
  env                     = nullptr;
  actions_count           = 0;

  hyperparameters.alpha   = 0.0;
  hyperparameters.gamma   = 0.0;
  hyperparameters.epsilon = 0.0;
  hyperparameters.epsilon_best = 0.0;

  run_best      = false;
}

Agent::Agent(Agent& other)
{
  copy(other);
}

Agent::Agent(const Agent& other)
{
  copy(other);
}

Agent::Agent(Env &env, sAgentHyperparameters hyperparameters)
{
  init(env, hyperparameters);
}

Agent::Agent(const Env &env, sAgentHyperparameters hyperparameters)
{
  init(env, hyperparameters);
}

Agent::~Agent()
{

}

Agent& Agent::operator= (Agent& other)
{
  copy(other);
  return *this;
}

Agent& Agent::operator= (const Agent& other)
{
  copy(other);
  return *this;
}


void Agent::init(Env &env, sAgentHyperparameters hyperparameters)
{
  this->env             = &env;
  this->hyperparameters = hyperparameters;
  this->actions_count   = this->env->get_actions_count();
  this->run_best      = false;
}

void Agent::init(const Env &env, sAgentHyperparameters hyperparameters)
{
  this->env             = (Env*)&env;
  this->hyperparameters = hyperparameters;
  this->actions_count   = this->env->get_actions_count();
  this->run_best      = false;
}

void Agent::copy(Agent &other)
{
  env               = other.env;
  actions_count     = other.actions_count;
  hyperparameters   = other.hyperparameters;
  run_best          = other.run_best;
}

void Agent::copy(const Agent &other)
{
  env               = other.env;
  actions_count     = other.actions_count;
  hyperparameters   = other.hyperparameters;
  run_best          = other.run_best;
}

void Agent::main()
{
  unsigned int action_id = rand()%actions_count;
  env->action(action_id);
}



unsigned int Agent::select_action(std::vector<float> &q_values)
{
  unsigned int result = 0;

  for (unsigned int i = 0; i < q_values.size(); i++)
    if (q_values[i] > q_values[result])
      result = i;

  float epsilon;
  if (run_best == true)
    epsilon = hyperparameters.epsilon_best;
  else
    epsilon = hyperparameters.epsilon;

  float v = (rand()%100000)/100000.0;
  if (v < epsilon)
    result = rand()%q_values.size();

  return result;
}
 
float Agent::max(std::vector<float> &v)
{
  unsigned int max_i = 0;

  for (unsigned int i = 0; i < v.size(); i++)
    if (v[i] > v[max_i])
      max_i = i;

  return v[max_i];
}
